### PR TITLE
Removed 'unit test' code from MessageListener. Cleaned up MetadataReader

### PR DESCRIPTION
**Pull request by dorkbox**
Not a lot, but moved code that was UnitTest only, into the unit test. Also, moved handles() checks to happen once, instead of 3x times.

### CODE REVIEW

**Comment by bennidi** [comment]
@dorkbox 
Looks like you have spent quite some work on your version of mbassador. I am happy to see that. Your pull request contains about twenty commits with a lot of code. I can not merge that reight now as it seems that you have removed some configuration functionality and also other parts of the public interface. But your code looks interesting. What kind of benefits do you get from using disruptor in this particular case? Do you have such a huge amount of messages to process? And what about asm reflection? Is that really so much faster?

Maybe you can take a look at my repo eventbus-performance and benchmark your version with the default mbassador?

**Comment by dorkbox** [comment]
Oh WHOOPS !


Originally, the pull request was going to be for just adding
multi-parameter methods/messages and the lock upgrade -- but I ran into
problems with "Enveloped" methods, and the pull request was not supposed to
include what it did *just some unit-test cleanup* - and when I made the PR,
I thought the pull request was for a specific commit -- not HEAD ! My
apologies.

Yes, I have HUGE numbers of messages (the goal is to replace a library I
use for networking)

 - lock-upgrades (from a readLock -> writeLock, safer DCL)  (
https://code.google.com/p/concurrent-locks/)
 - using the disruptor
<https://github.com/LMAX-Exchange/disruptor/wiki/Performance-Results>  is
significantly faster, also no need for extra objects to get created, more
specifically when calling publishAsync().
 - using reflectASM <https://github.com/EsotericSoftware/reflectasm> is
significantly faster

ReflectASM generates (and injects) bytecode that calls the method directly,
so (minus the bytecode generation part), it is *almost* as fast as direct
invocation (and significantly faster than reflection, even considering
inflation). Additionally, stack traces with reflectASM look a lot better.


I'll run some benchmarks VS the default mbassador, and let you know -- I
haven't run any performance benchmarks yet, as I've just been working on
just merging the features of the two projects.

**Comment by bennidi** [comment]
I understand that disruptor would reduce the synchronization overhead of the synchronized queues. Are you using it to replace the async message publication part only?
Regarding reflection, I am not so sure about the performance increase. JIT optimization is pretty efficient and I once benchmarked reflection (ReflectiveHandlerInvocation) with direct method invocation using a custom InvocationHandler implementation. Reflection was actually faster (suprisingly). I have not been able to fully make sense of it because there is a lot of contradicting talk out there about reflection performance and JIT optimizations. Another question is: Would reflectASM work on Dalvik?
And can you explain your ideas about the concurrent-locks implementation, regarding upgrade and all? Sounds interesting.

**Comment by dorkbox** [comment]
Yes, I'm using the disruptor exclusively for async message publication.

RE: reflection, JIT optimization is pretty good, and you're right, there is
a lot of contradicting talk about it, and given that I've worked a lot with
EsotericSoftware (really smart guy), I'm trusting his info. However, to
answer your question: ReflectASM doesn't work on Dalvik. Fortunately, it's
easy to take it out.

I'll test reflectASM/reflection, and compare the two -- and get back to
you. Your benchmark suite is nice. One question though -- does it spin-up
the JVM so it can settle?


Concurrent-locks isn't my impl, however, from what I understand, it
introduces a new lock, upgradeLock, which is permits readLocks, but blocks
writeLocks. If the lock is upgraded to a writeLock, it will block readLocks
and writeLocks. The idea is to remove the need for double-checked-locking.
Basically, it does what one would initially expect with locks, to go from a
readLock -> writeLock -> readLock.

**Comment by bennidi** [comment]
The benchmarking library does not yet include a ramp-up phase - actually I
wonder why. Should be very easy to build it in terms of a workload that is
run before everything else.  I will try to have a look into it tomorrow and
see whether I can find a quick solution. Unfortunately I am quite busy with
my work on OpenMed so I cannot put in much effort.

I think it will be very interesting to see the differences between your and
the my version.

After a quick look into the concurrent-locking library it seems to have the
same semantics as the lock handling in the SubscriptionManager. I was
trying to keep out external dependencies (which is why I built all this
ReflectionUtils crap instead of using a good library for it). Maybe its
time to give up this no dependency philosophy?!


I will tell you tomorrow how far I got regarding the benchmark suite.




2015-02-09 16:00 GMT+01:00 dorkbox <notifications@github.com>:

> Yes, I'm using the disruptor exclusively for async message publication.
>
> RE: reflection, JIT optimization is pretty good, and you're right, there is
> a lot of contradicting talk about it, and given that I've worked a lot with
> EsotericSoftware (really smart guy), I'm trusting his info. However, to
> answer your question: ReflectASM doesn't work on Dalvik. Fortunately, it's
> easy to take it out.
>
> I'll test reflectASM/reflection, and compare the two -- and get back to
> you. Your benchmark suite is nice. One question though -- does it spin-up
> the JVM so it can settle?
>
>
> Concurrent-locks isn't my impl, however, from what I understand, it
> introduces a new lock, upgradeLock, which is permits readLocks, but blocks
> writeLocks. If the lock is upgraded to a writeLock, it will block readLocks
> and writeLocks. The idea is to remove the need for double-checked-locking.
> Basically, it does what one would initially expect with locks, to go from a
> readLock -> writeLock -> readLock.
>
>
> On Mon, Feb 9, 2015 at 3:39 PM, Benjamin Diedrichsen <

**Comment by dorkbox** [comment]
Hm. I think that a "no dependency" solution, as much as possible, is a good
thing. There are too many libraries out there that are bloated.

I'll visit locking in the benchmarks as well, to see what impact they have.
I didn't investigate the similarities between it and what you did though.

**Comment by bennidi** [comment]
@dorkbox 
I thought about a configurable ramp up period and I remember why I did not include it. Firstly, benchmarks are intended to run for a longer time (it's not a micro benchmarking library) such that JIT will run and the JVM settles in the beginning of the run. This will be plotted in the graph as well such that you can actually see execution times dropping while the JIT compiler is doing its work. I think this is actually quite nice. Secondly, ramp up requirements might vary a lot between different scenarios. I think it is fairly easy to setup a workload as "ramp-up" work and make all the others start when it finished. This way warming up of the JVM does not need an extra concept in the library but is configurable with the basic mechanisms it provides. I think that is a better solution. But there might some sugar missing to do that elegantly. I would be happy to hear your opinion about it after you tried it yourself. Happy coding!

**Comment by dorkbox** [comment]
Good point on the ramp-up, and that this is not a micro benchmark library.
I agree and don't think it's necessary. Extra options to run a warmup
session would likely be good.

Here's some initial tests. I also included JVM memory usage, to help
understand the data. I don't think I have the disruptor completely worked
out, as it seems really, really slow. One thing I did notice is that the
default LBQ setting on MBassador used up to 1 GB of ram! Changing it to
have an LBQ of size 16 dropped it to about 512 MB without affecting too
much. The disruptor with a ring buffer of size 16 uses ~62 MB.

**Comment by bennidi** [comment]
You say "here are some initial tests". I ask, where ?! :)
There is no attachment or anything. Did you plot some graphs? Can you maybe post them in a new issue in the eventbus-performance repo? And reference it here?

**Comment by dorkbox** [comment]
It was attached - but I did this over email, which might explain why you
didn't see it.

I'll post them as an issue in the eventbus-performance repo.
